001-core/constants.php

```diff
@@ -0,0 +1,44 @@
+<?php
+
+namespace Automattic\VIP\Core\Constants;
+
+/**
+ * We use HyperDB, so DB_* constants are not automatically defined.
+ * We define them here in case any customer code is referencing them to avoid fatals errors in PHP 8+.
+ * As a best practice, these constants should not be used directly (`_doing_it_wrong()`).
+ *
+ * @param Object $wpdb
+ * @return void
+ */
+function vip_define_db_constants( $wpdb ) {
+	if ( defined( 'DB_NAME' ) || defined( 'DB_HOST' ) || defined( 'DB_PASSWORD' ) || defined( 'DB_USER' ) ) {
+		return;
```

Comment:
Could there ever be a situation where only some of these are defined?

I wonder if this function could be tweaked so that the early guard clause checks for at least one of them being undefined, and then adds a `defined()` wrapper around each of the `define()`'s at the end. That way, we fill the gaps, rather than do an all-or-nothing approach.

```suggestion
	if ( defined( 'DB_NAME' ) && defined( 'DB_HOST' ) && defined( 'DB_PASSWORD' ) && defined( 'DB_USER' ) ) {
	    // All DB_* constants are defined, so nothing else to do here.
		return;
```

then:

```
if ( ! defined( `DB_HOST` ) {
	define( 'DB_HOST', $db[0]['host'] );
}
```

etc.

Of course, the downside is that then the collection of constants may be inconsistent with each other in terms of a workable config, so maybe instead of just a return if only some of the constants are defined, there's an opportunity to capture / log that somewhere?

Reply:
> Could there ever be a situation where only some of these are defined?

This should not happen. "Normal" WordPress installations have all these constants defined. VIP installations have no constants defined because they use `$db_servers` and HyperDB.

In theory, a user on a "normal" installation can define only some of these constants; but if this is the case, they probably know what they are doing and why.

Reply:
>  That way, we fill the gaps, rather than do an all-or-nothing approach.

I prefer the all-or-nothing approach in this aspect because if only some are defined, then something is definitely off and the data being supplied from us may not be in tangent with what is already defined.